File: partitionequalsubsetsum.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        tr = sum(nums)
        if tr % 2 != 0:
            return False

        tr = tr/2
        t = {}

        def subsetSum(n, tr):
            if tr == 0:
                return True
            if n == 0:
                return False
            if (n, tr) in t:
                return t[(n, tr)]
            if nums[n-1] <= tr:
                t[(n, tr)] = subsetSum(n-1, tr-nums[n-1]) or subsetSum(n-1, tr)
                return t[(n, tr)]
            else:
                t[(n, tr)] = subsetSum(n-1, tr)
                return t[(n, tr)]

        return subsetSum(len(nums), tr)

File: Cpp/partitionequalsubsetsum.cpp
bool subsetsum(vector<int> &nums, int n, int target, vector<vector<int>> &memo) {
    if (target == 0) {
        return true;
    }
    if (n == 0) {
        return false;
    }
    if (memo[n][target] != -1) {
        return memo[n][target];
    }
    if (nums[n - 1] <= target) {
        memo[n][target] = subsetsum(nums, n - 1, target - nums[n - 1], memo) || subsetsum(nums, n - 1, target, memo);
        return memo[n][target];
    } else {
        memo[n][target] = subsetsum(nums, n - 1, target, memo);
        return memo[n][target];
    }
}
class Solution {
public:
    bool canPartition(vector<int> &nums) {
        int sum = 0;
        for (int i = 0; i < nums.size(); i++) {
            sum += nums[i];
        }
        if (sum % 2 != 0) {
            return false;
        }
        vector<vector<int>> memo(nums.size() + 1, vector<int>(sum / 2 + 1, -1));

        return subsetsum(nums, nums.size(), sum / 2, memo);
    }
};